<a name = Section1></a>
#### **1. Importing Libraries**

In [1]:
import pandas as pd                                                 # Importing for panel data analysis
#-------------------------------------------------------------------------------------------------------------------------------
import numpy as np                                                  # Importing package numpys (For Numerical Python)
#-------------------------------------------------------------------------------------------------------------------------------
import matplotlib.pyplot as plt                                     # Importing pyplot interface of matplotlib
import seaborn as sns                                               # Importing seaborn library for interactive visualization
%matplotlib inline
#-------------------------------------------------------------------------------------------------------------------------------
import warnings                                                     # Importing warning to disable runtime warnings
warnings.filterwarnings("ignore")                                   # Warnings will appear only once

<a name = Section1></a>
#### **2. Data Acquisition and Description**

Lets analyze the dataset and identify what attributes require generalization/categorization before we perform BUC on them.

In [2]:
data = pd.read_excel('master.xlsx') # Load the Excel dataset
print('Shape of the dataset:', data.shape)
data.head(3)

Shape of the dataset: (27820, 12)


country  year     sex          age  suicides_no  population  \
0  Albania  1987    male  15-24 years           21      312900   
1  Albania  1987    male  35-54 years           16      308000   
2  Albania  1987  female  15-24 years           14      289700   

   suicides/100k pop country-year  HDI for year   gdp_for_year ($)   \
0               6.71  Albania1987           NaN          2156624900   
1               5.19  Albania1987           NaN          2156624900   
2               4.83  Albania1987           NaN          2156624900   

   gdp_per_capita ($)    generation  
0                 796  Generation X  
1                 796        Silent  
2                 796  Generation X

- We have 27820 records and 12 attributes.
- In our records, we have variety of data including nominal data, binomial data, numerical data.

In [3]:
data.info() # Display basic information about the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27820 entries, 0 to 27819
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             27820 non-null  object 
 1   year                27820 non-null  int64  
 2   sex                 27820 non-null  object 
 3   age                 27820 non-null  object 
 4   suicides_no         27820 non-null  int64  
 5   population          27820 non-null  int64  
 6   suicides/100k pop   27820 non-null  float64
 7   country-year        27820 non-null  object 
 8   HDI for year        8364 non-null   float64
 9    gdp_for_year ($)   27820 non-null  int64  
 10  gdp_per_capita ($)  27820 non-null  int64  
 11  generation          27820 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.5+ MB


- Here we show the data type of our various attributes.

Always check for:
1. duplicate values in rows - delete duplicate rows
2. missing values in column

In [4]:
duplicate = data[data.duplicated()] # Selecting duplicate rows except first occurrence based on all columns
duplicate

Empty DataFrame
Columns: [country, year, sex, age, suicides_no, population, suicides/100k pop, country-year, HDI for year,  gdp_for_year ($) , gdp_per_capita ($), generation]
Index: []

- It means there are no duplicate records

In [5]:
print(data.isnull().sum())  # Check for missing values

country                   0
year                      0
sex                       0
age                       0
suicides_no               0
population                0
suicides/100k pop         0
country-year              0
HDI for year          19456
 gdp_for_year ($)         0
gdp_per_capita ($)        0
generation                0
dtype: int64


- As we can observe HDI has got 19,456 null values, out of total 27,820 entries. Given, more than half of the entries having NULL values, let's discount this column.

In [6]:
data.drop(['HDI for year'], axis=1, inplace=True)   # Remove the mentioned column

In [7]:
data.columns # remaining columns in our dataframe

Index(['country', 'year', 'sex', 'age', 'suicides_no', 'population',
       'suicides/100k pop', 'country-year', ' gdp_for_year ($) ',
       'gdp_per_capita ($)', 'generation'],
      dtype='object')

<a name = Section2></a>
#### **3. Data Analysis and AOI**

Now, let's one by one, analyze the 11 dimensions and determine for which dimension, we need to perform Attribute Oriented Induction (AOI) for generalization/categorization.

Data generalization summarizes data by replacing relatively low-level values with higher-level concepts, or by reducing the number of dimensions to summarize data in concept space involving fewer dimensions.

In [8]:
print("Unique values in coloumn country:\n", data["country"].unique())
print("---------------------------------------------------------")
print("Number of unique values:", data["country"].nunique())

Unique values in coloumn country:
 ['Albania' 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Aruba' 'Australia'
 'Austria' 'Azerbaijan' 'Bahamas' 'Bahrain' 'Barbados' 'Belarus' 'Belgium'
 'Belize' 'Bosnia and Herzegovina' 'Brazil' 'Bulgaria' 'Cabo Verde'
 'Canada' 'Chile' 'Colombia' 'Costa Rica' 'Croatia' 'Cuba' 'Cyprus'
 'Czech Republic' 'Denmark' 'Dominica' 'Ecuador' 'El Salvador' 'Estonia'
 'Fiji' 'Finland' 'France' 'Georgia' 'Germany' 'Greece' 'Grenada'
 'Guatemala' 'Guyana' 'Hungary' 'Iceland' 'Ireland' 'Israel' 'Italy'
 'Jamaica' 'Japan' 'Kazakhstan' 'Kiribati' 'Kuwait' 'Kyrgyzstan' 'Latvia'
 'Lithuania' 'Luxembourg' 'Macau' 'Maldives' 'Malta' 'Mauritius' 'Mexico'
 'Mongolia' 'Montenegro' 'Netherlands' 'New Zealand' 'Nicaragua' 'Norway'
 'Oman' 'Panama' 'Paraguay' 'Philippines' 'Poland' 'Portugal'
 'Puerto Rico' 'Qatar' 'Republic of Korea' 'Romania' 'Russian Federation'
 'Saint Kitts and Nevis' 'Saint Lucia' 'Saint Vincent and Grenadines'
 'San Marino' 'Serbia' 'Seychelles' 'Singap

- We will use the values of 'country' dimension as it is, because it is already in the highest-level of concept hierarchy.

In [9]:
print("Unique values in coloumn year:\n", data["year"].unique())
print("--------------------------------------------")
print("Number of unique values:", data["year"].nunique())

Unique values in coloumn year:
 [1987 1988 1989 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002
 2003 2004 2005 2006 2007 2008 2009 2010 1985 1986 1990 1991 2012 2013
 2014 2015 2011 2016]
--------------------------------------------
Number of unique values: 32


- We will use the values of 'year' dimension as it has 32 distinct values.

In [10]:
print("Unique values in coloumn sex:\n", data["sex"].unique())
print("--------------------------------------------")
print("Number of unique values:", data["sex"].nunique())

Unique values in coloumn sex:
 ['male' 'female']
--------------------------------------------
Number of unique values: 2


- We will use the values of 'sex' attribute/dimension as it is because it is already generalized, having two distinct values of the dimension.

In [11]:
print("Unique values in coloumn age:\n", data["age"].unique())
print("--------------------------------------------")
print("Number of unique values:", data["age"].nunique())

Unique values in coloumn age:
 ['15-24 years' '35-54 years' '75+ years' '25-34 years' '55-74 years'
 '5-14 years']
--------------------------------------------
Number of unique values: 6


- We will use the values of 'age' dimension as it is because it is already characterized by six distinct values of the dimension.

In [12]:
print("Unique values in coloumn suicide_no:\n", data["suicides_no"].unique())
print("--------------------------------------------")
print("Number of unique values:", data["suicides_no"].nunique())

Unique values in coloumn suicide_no:
 [  21   16   14 ... 5503 4359 2872]
--------------------------------------------
Number of unique values: 2084


- "suicide_no" can't be used directly. Need to perform AOI to create higher-level descriptions or categories for numerical data.

In [13]:
data["suicides_no"].describe()    # describe the values of the suicides_no attribute. 

count    27820.000000
mean       242.574407
std        902.047917
min          0.000000
25%          3.000000
50%         25.000000
75%        131.000000
max      22338.000000
Name: suicides_no, dtype: float64

In [14]:
data["suicides_no"].value_counts()

suicides_no
0       4281
1       1539
2       1102
3        867
4        696
        ... 
2158       1
525        1
2297       1
5241       1
2872       1
Name: count, Length: 2084, dtype: int64

- We can derive from the describe() that, the minimum number of suicides are 0. The maximum number of suicide value is 22338. 
At 25th percentile, the suicide value is 3. This means that 25 percent of data that lies below this 25th percentile point will have value equal to or less than 3. 
- At 50th percentile, the suicide value is 25. This means half of the data points below 50th percentile point will have value equal to or less than 25. For the high-level description purpose, we can label all those values as `low_suicide_range`.
- At 75th percentile, the suicide value is 131. This means that 75% of the data points that lies below this 75th percentile point will have value equal to or less than 131. For the high-level description purpose, we can label all those values above `low_suicide_range` and below the value at 75th percentile as `medium_suicide_range`.
- Similarly, the maximum suicide number reported is 22338. All values that lie between 75th percentile value to the maximum reported value can be termed as `high_suicide_range`.

In [15]:
low_suicides_range = data[(data["suicides_no"] >= data["suicides_no"].min()) & (data["suicides_no"] <= data["suicides_no"].quantile(0.50))]
medium_suicides_range = data[(data["suicides_no"] > data["suicides_no"].quantile(0.50)) & (data["suicides_no"] <= data["suicides_no"].quantile(0.75))]
high_suicides_range = data[(data["suicides_no"] > data["suicides_no"].quantile(0.75)) & (data["suicides_no"] <= data["suicides_no"].max())]

In [16]:
# Define the labels and conditions
conditions = [
    (data['suicides_no'] <= data['suicides_no'].quantile(0.5)),
    (data['suicides_no'] > data['suicides_no'].quantile(0.5)) & (data['suicides_no'] <= data['suicides_no'].quantile(0.75)),
    (data['suicides_no'] > data['suicides_no'].quantile(0.75))
]

labels = ['low_suicides_range', 'medium_suicides_range', 'high_suicides_range']

# Create a new column with the labels
data['suicides_range'] = np.select(conditions, labels, default='unknown')
data.drop(['suicides_no'], axis=1, inplace=True)   # Remove the column 'suicides_no' because we are using 'suicides_range' in place of that.
# Display the first few rows of the DataFrame with the new column
data.head(2)

country  year   sex          age  population  suicides/100k pop  \
0  Albania  1987  male  15-24 years      312900               6.71   
1  Albania  1987  male  35-54 years      308000               5.19   

  country-year   gdp_for_year ($)   gdp_per_capita ($)    generation  \
0  Albania1987          2156624900                 796  Generation X   
1  Albania1987          2156624900                 796        Silent   

       suicides_range  
0  low_suicides_range  
1  low_suicides_range

In [17]:
print("Unique values in column population:\n", data["population"].unique())
print("--------------------------------------------")
print("Number of unique values:", data["population"].nunique())

Unique values in column population:
 [ 312900  308000  289700 ... 2762158 2631600 1438935]
--------------------------------------------
Number of unique values: 25564


- "population" can't be used directly. Need to perform AOI to create higher-level descriptions or categories for numerical data.

population can't be used directly. Need to introduce some AOI here. Need to either generalize it or remove it completely. Can use histogram smoothing though.

In [18]:
data["population"].describe()    # describe the values of the population attribute. 

count    2.782000e+04
mean     1.844794e+06
std      3.911779e+06
min      2.780000e+02
25%      9.749850e+04
50%      4.301500e+05
75%      1.486143e+06
max      4.380521e+07
Name: population, dtype: float64

In [19]:
data["population"].describe().apply(lambda x: format(x, 'f')) # Suppress Scientific Notation

count       27820.000000
mean      1844793.617398
std       3911779.441756
min           278.000000
25%         97498.500000
50%        430150.000000
75%       1486143.250000
max      43805214.000000
Name: population, dtype: object

In [20]:
data["population"].value_counts()

population
24000      20
26900      13
20700      12
22000      12
4900       11
           ..
3282478     1
3953119     1
5745824     1
8448839     1
1438935     1
Name: count, Length: 25564, dtype: int64

- We can derive from the describe() that, the minimum reported population is 278. The maximum reported population is 43805214.
- At 25th percentile, the reported population value is 97498.5. This means that 25 percent of data that lies below this 25th percentile point will have value equal to or less than 97498. 
- At 50th percentile, the population value is 430150. This means half of the data points below 50th percentile point will have value equal to or less than 430150.
- At 75th percentile, the population value is 1486143.25. This means that 75% of the data points that lies below this 75th percentile point will have value equal to or less than 1486143.
- For the high-level description purpose:
    - we can label all those values that lie between 0 and 25th percentile value as `low_population_range`.
    - all the values that lie between 25th percentile value and 75th percentile value as `medium_population_range`.
    - all values that lie between 75th percentile value to the maximum reported value can be termed as `high_population_range`.

In [21]:
# Define the labels and conditions
conditions = [
    (data['population'] <= data['population'].quantile(0.25)),
    (data['population'] > data['population'].quantile(0.25)) & (data['population'] <= data['population'].quantile(0.75)),
    (data['population'] > data['population'].quantile(0.75))
]

labels = ['low_population_range', 'medium_population_range', 'high_population_range']

# Create a new column with the labels
data['population_range'] = np.select(conditions, labels, default='unknown')
data.drop(['population'], axis=1, inplace=True)   # Remove the column 'population' because we are using 'population_range' in place of that.
# Display the first few rows of the DataFrame with the new column
data.head(2)

country  year   sex          age  suicides/100k pop country-year  \
0  Albania  1987  male  15-24 years               6.71  Albania1987   
1  Albania  1987  male  35-54 years               5.19  Albania1987   

    gdp_for_year ($)   gdp_per_capita ($)    generation      suicides_range  \
0          2156624900                 796  Generation X  low_suicides_range   
1          2156624900                 796        Silent  low_suicides_range   

          population_range  
0  medium_population_range  
1  medium_population_range

In [22]:
print(data["suicides/100k pop"].value_counts())

suicides/100k pop
0.00     4281
0.29       72
0.32       69
0.34       55
0.37       52
         ... 
46.73       1
41.47       1
61.03       1
28.25       1
26.61       1
Name: count, Length: 5298, dtype: int64


In [23]:
print("Unique values in the column \"suicides/100k pop\":\n", data["suicides/100k pop"].unique())
print("Number of unique values:", data["suicides/100k pop"].nunique())
print("--------------------------------------------")
print("Unique values in the column \"country-year\":\n", data["country-year"].unique())
print("Number of unique values:", data["country-year"].nunique())
print("--------------------------------------------")

Unique values in the column "suicides/100k pop":
 [ 6.71  5.19  4.83 ... 47.86 40.75 26.61]
Number of unique values: 5298
--------------------------------------------
Unique values in the column "country-year":
 ['Albania1987' 'Albania1988' 'Albania1989' ... 'Uzbekistan2012'
 'Uzbekistan2013' 'Uzbekistan2014']
Number of unique values: 2321
--------------------------------------------


- As we are already using 'population' and 'suicides_no' attributes in their generalized form, we can remove the dimension "suicides/100k pop" from our dataset.
- Similarly, we are using distinct values in 'country' and 'year' dimension, therefore, we will drop the dimension 'country-year'.

In [24]:
data.drop(['suicides/100k pop'], axis=1, inplace=True)   # Remove the mentioned column
data.drop(['country-year'], axis=1, inplace=True)   # Remove the mentioned column

In [25]:
data.head(2)

country  year   sex          age   gdp_for_year ($)   gdp_per_capita ($)  \
0  Albania  1987  male  15-24 years          2156624900                 796   
1  Albania  1987  male  35-54 years          2156624900                 796   

     generation      suicides_range         population_range  
0  Generation X  low_suicides_range  medium_population_range  
1        Silent  low_suicides_range  medium_population_range

In [26]:
data[" gdp_for_year ($) "].describe()

count    2.782000e+04
mean     4.455810e+11
std      1.453610e+12
min      4.691962e+07
25%      8.985353e+09
50%      4.811469e+10
75%      2.602024e+11
max      1.812071e+13
Name:  gdp_for_year ($) , dtype: float64

In [27]:
data[" gdp_for_year ($) "].describe().apply(lambda x: format(x, 'f')) # Suppress Scientific Notation

count             27820.000000
mean       445580969025.726624
std       1453609985940.912109
min            46919625.000000
25%          8985352832.000000
50%         48114688201.000000
75%        260202429150.000000
max      18120714000000.000000
Name:  gdp_for_year ($) , dtype: object

- We can derive from the describe() that, the minimum reported gdp_per_year value is 46,919,625 $. The maximum reported population is 18,120,714,000,000. 
- At 25th percentile, the reported gdp_per_year value is 8,985,352,832 $. This means that 25 percent of data that lies below this 25th percentile point will have value equal to or less than 8,985,352,832. 
- At 50th percentile, the gdp_per_year value is 48,114,688,201 $. This means half of the data points below 50th percentile point will have value equal to or less than 48,114,688,201.
- At 75th percentile, the gdp_per_year value is 260,202,429,150 $. This means that 75% of the data points that lies below this 75th percentile point will have value equal to or less than 260,202,429,150.
- For the high-level description purpose:
    - we can label all those values that lie between 0 and 25th percentile value as `low_income_range`.
    - all the values that lie between 25th percentile value and 75th percentile value as `medium_income_range`.
    - all values that lie between 75th percentile value to the maximum reported value can be termed as `high_income_range`.

In [28]:
# Define the labels and conditions
conditions = [
    (data[' gdp_for_year ($) '] <= data[' gdp_for_year ($) '].quantile(0.25)),
    (data[' gdp_for_year ($) '] > data[' gdp_for_year ($) '].quantile(0.25)) & (data[' gdp_for_year ($) '] <= data[' gdp_for_year ($) '].quantile(0.75)),
    (data[' gdp_for_year ($) '] > data[' gdp_for_year ($) '].quantile(0.75))
]

labels = ['low_income_range', 'medium_income_range', 'high_income_range']

# Create a new column with the labels
data['gdp_per_year_income_range'] = np.select(conditions, labels, default='unknown')
data.drop([' gdp_for_year ($) '], axis=1, inplace=True)   # Remove the column 'gdp_per_year ($)' because we are using 'gdp_per_year_income_range' in place of that.
# Display the first few rows of the DataFrame with the new column
data.head(2)

country  year   sex          age  gdp_per_capita ($)    generation  \
0  Albania  1987  male  15-24 years                 796  Generation X   
1  Albania  1987  male  35-54 years                 796        Silent   

       suicides_range         population_range gdp_per_year_income_range  
0  low_suicides_range  medium_population_range          low_income_range  
1  low_suicides_range  medium_population_range          low_income_range

In [29]:
data["gdp_per_capita ($)"].describe()

count     27820.000000
mean      16866.464414
std       18887.576472
min         251.000000
25%        3447.000000
50%        9372.000000
75%       24874.000000
max      126352.000000
Name: gdp_per_capita ($), dtype: float64

In [30]:
bins = [i*2000 for i in range(int(data['gdp_per_capita ($)'].quantile(0.75))//1000)]
bins.append(int(data['gdp_per_capita ($)'].max()))
labels = ["{}-{}".format(bins[i], bins[i+1]) for i in range(len(bins) - 1)]
print(bins)
labels[-1]=">{}".format(bins[-2])
print(labels)

data["gdp_per_capita_range"] = pd.cut(data["gdp_per_capita ($)"], bins=bins, labels=labels)

data.drop(['gdp_per_capita ($)'], axis=1, inplace=True)   # Remove the column 'gdp_per_capita ($)' because we are using 'gdp_per_capita_range' in place of that.
# Display the first few rows of the DataFrame with the new column
data.head(2)

[0, 2000, 4000, 6000, 8000, 10000, 12000, 14000, 16000, 18000, 20000, 22000, 24000, 26000, 28000, 30000, 32000, 34000, 36000, 38000, 40000, 42000, 44000, 46000, 126352]
['0-2000', '2000-4000', '4000-6000', '6000-8000', '8000-10000', '10000-12000', '12000-14000', '14000-16000', '16000-18000', '18000-20000', '20000-22000', '22000-24000', '24000-26000', '26000-28000', '28000-30000', '30000-32000', '32000-34000', '34000-36000', '36000-38000', '38000-40000', '40000-42000', '42000-44000', '44000-46000', '>46000']


country  year   sex          age    generation      suicides_range  \
0  Albania  1987  male  15-24 years  Generation X  low_suicides_range   
1  Albania  1987  male  35-54 years        Silent  low_suicides_range   

          population_range gdp_per_year_income_range gdp_per_capita_range  
0  medium_population_range          low_income_range               0-2000  
1  medium_population_range          low_income_range               0-2000

In [31]:
print(data["generation"].value_counts())

generation
Generation X       6408
Silent             6364
Millenials         5844
Boomers            4990
G.I. Generation    2744
Generation Z       1470
Name: count, dtype: int64


In [32]:
print("Unique values:\n", data["generation"].unique())
print("--------------------------------------------")
print("Number of unique values:", data["generation"].nunique())

Unique values:
 ['Generation X' 'Silent' 'G.I. Generation' 'Boomers' 'Millenials'
 'Generation Z']
--------------------------------------------
Number of unique values: 6


- We will use the values of 'generation' dimension as it is, because it is generalized into 6 distinct values.

----

In [35]:
data.head(2)

country  year   sex          age    generation      suicides_range  \
0  Albania  1987  male  15-24 years  Generation X  low_suicides_range   
1  Albania  1987  male  35-54 years        Silent  low_suicides_range   

          population_range gdp_per_year_income_range gdp_per_capita_range  
0  medium_population_range          low_income_range               0-2000  
1  medium_population_range          low_income_range               0-2000

In [36]:
data.to_csv('preprocessed.csv', header=True, index=False)

----